### PR TITLE
Lock workshop type 1 month out from summer workshop

### PR DESCRIPTION
**Pull request by TurnerRiley**
For summer workshops for CSP and CSA we want to make it so only workshop admins can update the “Is this a virtual workshop” within one month of the workshop in order to prevent people from changing this last minute without talking to Friday institute. This PR addresses this by disabling the virtual field under the given criteria and modifies the associated tooltip to say, “There is a less than a month until your workshop. Please contact [partner@code.org](mailto:partner@code.org) to update this setting.”

### Demo showing that the field is disabled only when those 3 criteria are met
| CSP or CSA? | Workshop admin? | Within one month of the workshop starting? | Should field be disabled? | Screenshot |
| ------------  | ------------  | ------------    | -----------   | ------------ |
| No | No | No | No | [SS](https://user-images.githubusercontent.com/56283563/230481180-8cd00d34-cef0-4d63-8017-60bf4051ab8c.JPG) |
| No | No | Yes | No | [SS](https://user-images.githubusercontent.com/56283563/230481206-77833231-ce59-4403-86f5-10ab3675efda.JPG) |
| No | Yes | No | No | [SS](https://user-images.githubusercontent.com/56283563/230481231-650a70c0-85e1-446b-9b7f-077bd2dab571.JPG) |
| No | Yes | Yes | No | [SS](https://user-images.githubusercontent.com/56283563/230481265-38202d2a-5d28-4c08-8263-ad041a4a2a8e.JPG) |
| Yes | No | No | No | [SS](https://user-images.githubusercontent.com/56283563/230481304-c1928ac5-1354-4fd0-8815-590af9738ef1.JPG) |
| Yes | No | Yes | **Yes** | [SS](https://user-images.githubusercontent.com/56283563/230481338-b597ae6b-d736-4cc6-a209-1817c32a5565.JPG) |
| Yes | Yes | No | No | [SS](https://user-images.githubusercontent.com/56283563/230481406-0b5617ba-edc2-4cef-aa04-a678fefd1a15.JPG) |
| Yes | Yes | Yes | No | [SS](https://user-images.githubusercontent.com/56283563/230481415-f3fd3896-832f-454a-a841-9726aa1eec64.JPG) |

#### Demo of data validation throwing error if 'virtual' is changed under the conditions when it shouldn't be able to (in this case, I made the date over a month later, switched the virtual setting, then switched the date back to within a month from starting)
update_validation

### New workshop edge case
In the event a user is creating a new workshop that starts within a month of them creating it, we want to default to "In-person" so that they have to reach out and work with us or the Friday Institute to make it virtual if that's the case.

#### Behavior demo
Tooltip demo: here

#### Demo of data validation throwing error if somehow 'virtual' is selected in this case (this is theoretically impossible so I had to temporarily disable the feature that limits which options are available to demo this, but still good to have in case of future changes that might not catch this)
create_validation

## Links
Main Jira ticket: [here](https://codedotorg.atlassian.net/jira/software/c/projects/ACQ/boards/64?modal=detail&selectedIssue=ACQ-332&assignee=60d62161f65054006980bd71)
Edge case Jira ticket: [here](https://codedotorg.atlassian.net/jira/software/c/projects/ACQ/boards/64?modal=detail&selectedIssue=ACQ-507&assignee=60d62161f65054006980bd71)

## Testing story
Local testing and adding unit tests.

## PR Checklist:
- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx [hunk 1]
```diff
@@ -361,6 +377,30 @@ export class WorkshopForm extends React.Component {
...
+  workshopStartsWithinMonth(session) {
+    const today = new Date();
+    const workshopDate = new Date(session.date);
+    const monthBeforeWorkshopDate = new Date(
+      workshopDate.getFullYear(),
+      workshopDate.getMonth() - 1,
+      workshopDate.getDate()
+    );
+    return monthBeforeWorkshopDate <= today && today <= workshopDate;
+  }
```
Comment (TurnerRiley) [inline apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx]: I originally had this as a function in `workshop_serializer.rb` to quickly determine the start date of a workshop so I wouldn't have to iterate through the sessions to check but I ran into 2 issues. 1) In `workshop.rb`, the `workshop_starting_date` is determined by which session is listed first not which comes first chronologically (which probably wouldn't be a big deal since most people would probably list the first session first, but just in case). 2) When a user is first creating a workshop there is no workshop entry yet to see if the sessions are within a month or not, so I would have to check the ones listed in this `workshop_form` anyways. So instead of having separate logic for editing an existing workshop vs. creating a new one, I figured it would be simpler to just use the same process for both here.

File: apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx [hunk 2]
```diff
@@ -419,16 +467,19 @@ export class WorkshopForm extends React.Component {
...
                 onChange={this.handleVirtualChange}
                 readOnly={
                   this.props.readOnly ||
-                  VirtualOnlySubjects.includes(this.state.subject)
+                  VirtualOnlySubjects.includes(this.state.subject) ||
+                  this.state.cannotChangeIfWorkshopVirtual
+                }
+                showVirtualOptions={
+                  !!this.props.workshop ||
+                  !this.state.cannotChangeIfWorkshopVirtual
```
Comment (TurnerRiley) [inline apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx]: This is for the edge case if someone is creating a workshop and it is within a month of the start date. In this case, 'in-person' will be the only option to a non-admin.

**Comment by davidsbailey** [comment]
This looks like a good step, but when I am signed in as an RP program manager, and i edit a workshop that is starting soon, it still looks easy to change it to virtual by changing the workshop subject:

https://user-images.githubusercontent.com/8001765/232608573-412d04dd-e642-4819-9206-a57596c22b09.mov

I think the solution may require a bit of PM input -- can we just freeze the workshop subject and course menus as well? does the behavior need to be different for "new" mode vs "edit" mode? etc.

**Comment by TurnerRiley** [comment]
> This looks like a good step, but when I am signed in as an RP program manager, and i edit a workshop that is starting soon, it still looks easy to change it to virtual by changing the workshop subject:
> 
>  Screen.Recording.2023-04-17.at.1.53.44.PM.mov 
> I think the solution may require a bit of PM input -- can we just freeze the workshop subject and course menus as well? does the behavior need to be different for "new" mode vs "edit" mode? etc.

Good point! I'll connect with Dani and report back with the desired behavior.

**Comment by TurnerRiley** [comment]
> This looks like a good step, but when I am signed in as an RP program manager, and i edit a workshop that is starting soon, it still looks easy to change it to virtual by changing the workshop subject:
> 
>  Screen.Recording.2023-04-17.at.1.53.44.PM.mov 
> I think the solution may require a bit of PM input -- can we just freeze the workshop subject and course menus as well? does the behavior need to be different for "new" mode vs "edit" mode? etc.

I checked in with Dani and she says it's fine to not cover that case since we are more prioritizing deterring a user from changing the virtual-ness of a section, rather than covering all ways they could change it (since they are already instructed not to do so when it's close to the start of a workshop).

File: apps/test/unit/code-studio/pd/workshop_dashboard/workshopFactory.js [hunk 3]
```diff
@@ -33,14 +38,74 @@ Factory.define('csp summer workshop')
...
+    ).toISOString()
+  )
+  .attr(
+    'end',
+    new Date(
+      today.getFullYear(),
+      today.getMonth() + 2,
+      today.getDate()
+    ).toISOString()
+  );
```
Comment (davidsbailey) [inline apps/test/unit/code-studio/pd/workshop_dashboard/workshopFactory.js:184]: nice job adding these tests! one concern about using `today` is that it can cause our tests to behave differently depending on the current date/time, which is something we want to avoid. could you please change these to us a specific date of your choice, and then also include tests showing that the "over a month" logic still works across the end-of-year boundary?

Comment (TurnerRiley) [reply apps/test/unit/code-studio/pd/workshop_dashboard/workshopFactory.js:184]: Since it has to check if the workshop is within a month, the logic uses `today`, is it possible for me to set an arbitrary "today" like with TimeCop? And yes I can also add that!

Comment (davidsbailey) [reply apps/test/unit/code-studio/pd/workshop_dashboard/workshopFactory.js:184]: ah I see, this is tricky! here are a couple ideas:
1. extract a helper method like [this one](https://github.com/code-dot-org/code-dot-org/blob/54f5a314e5c812e03a008afed2127d6080601f04/apps/src/utils.js#L739-L741) to fetch the current date, which you can then stub out in your unit tests (like [here](https://github.com/code-dot-org/code-dot-org/blob/563e25389306aa2f0ef2dd8e72c7198da73edb1d/apps/test/unit/code-studio/initApp/projectTest.js#L743))
2. do the date calculation on the server. this is a slightly bigger change, but will be more reliable since browser date/time is not reliable anyway.

if you are feeling ambitious then (2) would be best, but since you've already got most of this working and (per other discussion topic) the quality bar seems to be "best effort" right now, then I think (1) is also fine and could avoid undoing a lot of your existing work and testing.

one thing to call out is that if we do go with (1), there's the risk that someone with a wrong time on their machine won't be able to edit the "virtual" field on their workshop even if it is more than a month out, but since by definition they'll have over a month before their workshop in that scenario it seems like this risk is unlikely to be super disruptive (i.e. they will have time to contact us well ahead of their workshop).

Comment (TurnerRiley) [reply apps/test/unit/code-studio/pd/workshop_dashboard/workshopFactory.js:184]: I'm down to look into option 2! Would you mind explaining what you mean by "do the date calculation on the server" please? Like would that entail obtaining the date on the ruby side and passing that over to the workshop form to use there? Or am I misunderstanding this?

Comment (davidsbailey) [reply apps/test/unit/code-studio/pd/workshop_dashboard/workshopFactory.js:184]: sorry to leave out important details! at the very least, today's date should be be calculated on the server. but unless you need it for anything else, I would recommend calculating `workshopStartsWithinMonth` on the server, and passing that down as a boolean, in which case you won't need to pass down today's date. that will let you keep some of the logic you have on the client.

Comment (TurnerRiley) [reply apps/test/unit/code-studio/pd/workshop_dashboard/workshopFactory.js:184]: No worries at all! That's what I originally had but had to switch since when a user is creating a workshop, the workshop db entry isn't added until they submit the form so I wouldn't be able to access the dates of the workshop on the backend to check if they're starting within a month. That's why I moved it all to the front-end so that I can access the current values they have entered, plus this unifies the behavior regardless of if they're creating a new workshop or editing an existing one. So I'm thinking maybe I'll just calculate today's date on teh server-side and pass that in but keep the front-end logic? Curious on your thoughts on that!

Comment (davidsbailey) [reply apps/test/unit/code-studio/pd/workshop_dashboard/workshopFactory.js:184]: > So I'm thinking maybe I'll just calculate today's date on teh server-side and pass that in but keep the front-end logic? Curious on your thoughts on that!

ah, good point! yes, this makes sense to me!

File: apps/src/code-studio/pd/workshop_dashboard/DetailsPanel.jsx [hunk 4]
```diff
@@ -28,6 +28,10 @@ export default class DetailsPanel extends React.Component {
     showAdminEditConfirmation: false
   };
 
+  getToday = () => {
+    return new Date();
+  };
+
```
Comment (TurnerRiley) [inline apps/src/code-studio/pd/workshop_dashboard/DetailsPanel.jsx:34]: I put the declaration of "today" as a prop for WorkshopForm so that the today object is not re-calculated every single time there's a relevant update since that seems unnecessary (plus it simplifies setting up the tests).

File: apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx [hunk 5]
```diff
@@ -361,6 +379,32 @@ export class WorkshopForm extends React.Component {
...
+  // Returns whether today is within a month before the workshop start.
+  workshopStartsWithinMonth = session => {
+    const today = this.props.today;
+    const workshopDate = new Date(session.date);
+    const monthBeforeWorkshopDate = new Date(
+      workshopDate.getFullYear(),
+      workshopDate.getMonth() - 1,
+      workshopDate.getDate()
+    );
+    return monthBeforeWorkshopDate <= today && today <= workshopDate;
```
Comment (davidsbailey) [inline apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx:390]: wow, TIL that `new Date()` works properly with a negative number for month:

Screenshot 2023-04-24 at 10 42 11 AM

File: apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx [hunk 6]
```diff
@@ -361,6 +379,32 @@ export class WorkshopForm extends React.Component {
...
+      (course === ActiveCourseWorkshops.CSP ||
+        course === ActiveCourseWorkshops.CSA) &&
+      subject === SubjectNames.SUBJECT_SUMMER_WORKSHOP &&
+      !this.props.permission.has(WorkshopAdmin) &&
+      sessions.some(this.workshopStartsWithinMonth)
+    );
+  }
+
+  // Returns whether today is within a month before the workshop start.
+  workshopStartsWithinMonth = session => {
```
Comment (davidsbailey) [inline apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx]: I would recommend renaming this to `sessionStartsWithinMonth`

File: apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx [hunk 7]
```diff
@@ -725,6 +780,12 @@ export class WorkshopForm extends React.Component {
 
   handleCourseChange = event => {
     const course = this.handleFieldChange(event);
+    const cannotChangeIfWorkshopVirtual =
+      this.checkCannotChangeIfWorkshopVirtual(
+        this.state.sessions,
+        course,
+        this.state.subject
+      );
```
Comment (davidsbailey) [inline apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx]: heads up that this is leading to some weird behavior: 

https://user-images.githubusercontent.com/8001765/234135048-cd2cdf09-dcf4-4bce-b507-5bdca51b38c6.mov

you're using the new course value but the old subject value to compute the result. in this case it is probably harmless, because the problem is fixed after you select a subject, but please keep this kind of issue in mind for the future.

better alternatives for next time include:
* avoid duplicate state entirely by calling checkCannotChangeIfWorkshopVirtual only when you need it
* set a constant equal to the result of checkCannotChangeIfWorkshopVirtual at the beginning of the render method, without assigning it to component state
* if you must use state, consider using [setState callbacks](https://legacy.reactjs.org/docs/react-component.html#setstate) to ensure that all state updates have finished before relying on this.state (unsure what the equivalent pattern is using react hooks)

File: apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js [hunk 8]
```diff
@@ -63,4 +75,165 @@ describe('WorkshopForm test', () => {
...
+            today={getFakeToday(true)}
+            readOnly={false}
+          />
+        </MemoryRouter>
+      </Provider>
+    );
+
+    const virtualFormController = wrapper.find('#virtual').first();
+    assert(!virtualFormController.props().disabled);
+  });
```
Comment (davidsbailey) [inline apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js:238]: great tests!

File: apps/test/unit/code-studio/pd/workshop_dashboard/workshopFactory.js [hunk 9]
```diff
@@ -33,14 +42,158 @@ Factory.define('csp summer workshop')
...
+  .extend('workshop')
+  .attr('course', 'CS Principles')
+  .attr('subject', 'Academic Year Workshop 1')
+  .attr('sessions', () => Factory.buildList('session', 5))
+  .attr('account_required_for_attendance?', true)
+  .attr('scholarship_workshop?', true)
+  .attr('location_name', 'physical');
+
+Factory.define(
+  'csp summer workshop starting within month of middleOfYearFakeToday'
```
Comment (davidsbailey) [inline apps/test/unit/code-studio/pd/workshop_dashboard/workshopFactory.js]: I think you can safely shorten this up a bit, since this is text that's going to get reused a lot:
```suggestion
  'csp summer workshop starting within a month'
```
and just use the more specific text when near the end of the year

File: dashboard/test/controllers/api/v1/pd/workshops_controller_test.rb [hunk 10]
```diff
@@ -737,6 +776,49 @@ class Api::V1::Pd::WorkshopsControllerTest < ::ActionController::TestCase
...
+    assert_response :success
+  end
+
+  test 'updating virtual field in non-CSP/CSA summer workshop within a month of starting as a non-ws-admin does not raise error' do
+    sign_in @organizer
+    workshop = create :csd_summer_workshop, organizer: @organizer
+
+    put :update, params: {id: workshop.id, pd_workshop: workshop_params.merge(course: Pd::Workshop::COURSE_CSD, subject: Pd::Workshop::SUBJECT_CSD_SUMMER_WORKSHOP, funding_type: nil, virtual: true)}
+    assert_response :success
+  end
```
Comment (davidsbailey) [inline dashboard/test/controllers/api/v1/pd/workshops_controller_test.rb:820]: 💯

**Review by davidsbailey — COMMENTED** [review]
looking great Turner! Looks good to merge once this round of comments is addressed and tests are passing.

File: apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx (hunk 5 above)
Comment (TurnerRiley) [reply apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx:390]: Yes me too! I did a lot of playing around with the Date object to see how well it stood up to odd numbered entries (but it seems to resolve as expected each time)!

File: apps/test/unit/code-studio/pd/workshop_dashboard/workshopFactory.js (hunk 9 above)
Comment (TurnerRiley) [reply apps/test/unit/code-studio/pd/workshop_dashboard/workshopFactory.js]: Sounds good!

File: apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx (hunk 7 above)
Comment (TurnerRiley) [reply apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx]: Oh good point! That change was in the case that 'course' was being changed so that's why I used the newly-defined 'course' but the state-defined 'subject' and 'sessions'. You're right, there is no need for cannotChangeIfWorkshopVirtual in the state so I'll just remove it to keep things simpler and only call it when I need it! (I double checked after removing it from state and the behavior is the same and much less complicated!)

**Comment by TurnerRiley** [comment]
> looking great Turner! Looks good to merge once this round of comments is addressed and tests are passing.

Thank you! I think I made susbtantial enough changes that it removed your approval so that's why I re-requested review (just in case)!

**Review by davidsbailey — APPROVED** [review]
thank you, this cleaned up nicely!
